components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatsUpdateUtil.java

```diff
@@ -0,0 +1,100 @@
+
+import java.rmi.RemoteException;
+import java.util.Map;
+
+/**
+ * This class is used to initiate the stats publishing status in a distributed environment
+ */
+public class StatsUpdateUtil {
+
+    private static final Log log = LogFactory.getLog(StatsUpdateUtil.class);
+    public static StatsUpdateUtil instance = new StatsUpdateUtil();
+    private Map<String, Environment> gatewyEnvironments;
+
+    public static synchronized StatsUpdateUtil getInstance() {
+        if (instance == null) {
+            instance = new StatsUpdateUtil();
+        }
+        return instance;
+    }
+
+    /**
+     * This method is used to initiate the web service calls and cluster messages related to stats publishing status
+     *
+     * @param receiverUrl   event receiver url
+     * @param user          username of the event receiver
+     * @param password      password of the event receiver
+     * @param updatedStatus status of the stat publishing state
+     * @throws APIManagementException if an error occurs while trying to update status and send cluster message
+     */
+    public void callStatupdateService(String receiverUrl, String user, String password, boolean updatedStatus)
+            throws APIManagementException {
+
+        try {
+            gatewyEnvironments = ServiceReferenceHolder.getInstance().getAPIManagerConfigurationService().
+                    getAPIManagerConfiguration().getApiGatewayEnvironments();
+
+            for (String gatewayEnvironmentName : gatewyEnvironments.keySet()) {
+                Environment currentGatewayEnvironment = gatewyEnvironments.get(gatewayEnvironmentName);
+                String gatewayServiceUrl = currentGatewayEnvironment.getServerURL();
+                String gatewayUserName = currentGatewayEnvironment.getUserName();
+                String gatewayPassword = currentGatewayEnvironment.getPassword();
+
+                //get the stub and the call the admin service with the credentials
+                GatewayStatsUpdateServiceStub stub =
+                        new GatewayStatsUpdateServiceStub(gatewayServiceUrl + "GatewayStatsUpdateService");
+                StatUpdateStorePublisherDomain storePublisherMessageAgent = new StatUpdateStorePublisherDomain();
+                ServiceClient gatewayServiceClient = stub._getServiceClient();
+                CarbonUtils.setBasicAccessSecurityHeaders(gatewayUserName, gatewayPassword, gatewayServiceClient);
+
+                //send an empty string if at least one mandatory parameter is null
+                if (receiverUrl == null || user == null || password == null) {
+                    receiverUrl = "";
+                    user = "";
+                    password = "";
+                }
+                stub.updateStatPublishGateway(receiverUrl, user, password, updatedStatus);
+
+                //send cluster message to publisher-store domain
+                storePublisherMessageAgent.updateStatsPublishStore(updatedStatus);
+            }
```

Comment:
We should not throw exceptions here. If the message update fails for one Gateway, we should still continue to send the update to the other Gateways. Therefore we need to continue even on error and then provide a log of the failures later.